src/CoreFoundation/CFProxySupport.cs

```diff
@@ -579,62 +581,328 @@ public static CFProxySettings GetSystemProxySettings ()
+			IntPtr cbError = IntPtr.Zero;
+			if (proxyAutoConfigurationScript == null)
+				throw new ArgumentNullException (nameof (proxyAutoConfigurationScript));
 
-			if (clientContext == null)
-				throw new ArgumentNullException ("clientContext");
+			if (targetUrl == null)
+				throw new ArgumentNullException (nameof (targetUrl));
 
-			IntPtr source = CFNetworkExecuteProxyAutoConfigurationScript (proxyAutoConfigurationScript.Handle, targetURL.Handle, resultCallback, clientContext);
-			return (source == IntPtr.Zero) ? null : new CFRunLoopSource (source);
+			using (var pacScript = new NSString (proxyAutoConfigurationScript)) 
+			using (var url = new NSUrl (targetUrl.AbsoluteUri)) {
+				CreatePACCFRunLoopSource factory = delegate (CFProxyAutoConfigurationResultCallbackInternal cb, ref CFStreamClientContext context) {
+					return CFNetworkExecuteProxyAutoConfigurationScript (pacScript.Handle, url.Handle, cb, ref context);
+				};
+				return ExecutePacCFRunLoopSourceBlocking (factory, out outError);
+			}
 		}
 		
+		public static async Task<(CFProxy[] proxies, NSError error)> ExecuteProxyAutoConfigurationScriptAsync (string proxyAutoConfigurationScript, Uri targetUrl, CancellationToken cancellationToken)
+		{
+			if (proxyAutoConfigurationScript == null)
+				throw new ArgumentNullException (nameof (proxyAutoConfigurationScript));
+
+			if (targetUrl == null)
+				throw new ArgumentNullException (nameof (targetUrl));
+
+			using (var pacScript = new NSString (proxyAutoConfigurationScript)) 
+			using (var url = new NSUrl (targetUrl.AbsoluteUri)) {
+				CreatePACCFRunLoopSource factory = delegate (CFProxyAutoConfigurationResultCallbackInternal cb, ref CFStreamClientContext context) {
+					return CFNetworkExecuteProxyAutoConfigurationScript (pacScript.Handle, url.Handle, cb, ref context);
+				};
+				// use the helper task with a factory for this method
+				return await ExecutePacCFRunLoopSourceAsync (factory, cancellationToken).ConfigureAwait (false);
+			}
+		}
+
 		[DllImport (Constants.CFNetworkLibrary)]
 		extern static /* CFRunLoopSourceRef __nonnull */ IntPtr CFNetworkExecuteProxyAutoConfigurationURL (
 			/* CFURLRef __nonnull */ IntPtr proxyAutoConfigurationURL,
 			/* CFURLRef __nonnull */ IntPtr targetURL,
-			/* CFProxyAutoConfigurationResultCallback __nonnull */ IntPtr cb,
-			/* CFStreamClientContext * __nonnull */ IntPtr clientContext);
+			/* CFProxyAutoConfigurationResultCallback __nonnull */ CFProxyAutoConfigurationResultCallbackInternal cb,
+			/* CFStreamClientContext * __nonnull */ ref CFStreamClientContext clientContext);
 
-		public static CFRunLoopSource ExecuteProxyAutoConfigurationURL (NSUrl proxyAutoConfigurationURL, NSUrl targetURL, CFProxyAutoConfigurationResultCallback resultCallback, CFStreamClientContext clientContext)
+		public static CFRunLoopSource ExecuteProxyAutoConfigurationUrl (Uri proxyAutoConfigurationUrl, Uri targetUrl, CFProxyAutoConfigurationResultCallback clientCallback, NSObject clientContext)
```

Comment:
This API takes a `clientContext`, I suppose that's just anything the caller wants it to be (similar to the native API)? If so, it's redundant, since managed code can use anonymous delegates. It looks like just removing the `clientContext` parameter will make things somewhat simpler too.

Reply:
Removing it will certainly make my life easier, but I'm not sure if clients will want it. In a sense, less Marshalling etc.. so better for us making the code simpler. But this is the 'advance' API (The async/await one is much nicer). I have no preference to be hones (well, yes, I prefer to have less code).

Reply:
I tested, if the user use an anonymous delegate, we get an exception when marshalling :/

Reply:
Then you have a bug somewhere :) There's no reason we can't make anonymous delegates work as long as the native API provides a context parameter of some sort.

Reply:
Also it's not common in managed code to have a context parameter like this (people will wonder what the parameter is for), and we hide it everywhere else in our bindings too.